File: src/packages/arcane/lib/finding-lifecycle.mjs
// Finding lifecycle primitives operate on typed identities and evidence digests.
// They deliberately never inspect a finding's prose, line anchor, or narrative.
import { digestValue } from './canonical.mjs';
import { decision } from './errors.mjs';

const SEVERITY = Object.freeze({ informational: 0, low: 1, medium: 2, high: 3, critical: 4 });
const id = (value, camel, snake) => value?.[camel] ?? value?.[snake] ?? null;
const freeze = (value) => Object.freeze(value);

function required(value, name) {
  if (typeof value !== 'string' || !value) throw new TypeError(`${name} is required`);
  return value;
}

function findingIdentity(finding) {
  const controlId = id(finding, 'controlId', 'control_id') ?? id(finding, 'ruleId', 'rule_id');
  const subjectId = id(finding, 'subjectId', 'subject_id') ?? finding?.file ?? finding?.path;
  const semanticKey = id(finding, 'semanticKey', 'semantic_key')
    ?? id(finding, 'conditionFingerprint', 'condition_fingerprint') ?? null;
  required(controlId, 'finding controlId or ruleId');
  required(subjectId, 'finding subjectId');
  return semanticKey === null
    ? { control_id: controlId, subject_id: subjectId }
    : { control_id: controlId, subject_id: subjectId, semantic_key: semanticKey };
}

/** Stable identity for a finding. Presentation text & source line anchors are excluded. */
export function fingerprintFinding(finding) {
  return digestValue(findingIdentity(finding));
}

export const stableFindingFingerprint = fingerprintFinding;

function event(event_type, payload) {
  return freeze({ event_type, payload: freeze(payload) });
}

/**
 * Upsert a finding by its stable fingerprint. Existing records retain their
 * originating review round; only current structured observation changes.
 */
export function upsertFinding({ records = [], finding, reviewRoundId, review_round_id: reviewRoundIdSnake }) {
  const stable_fingerprint = fingerprintFinding(finding);
  const round = reviewRoundId ?? reviewRoundIdSnake;
  required(round, 'reviewRoundId');
  const existing = records.find((record) => record.stable_fingerprint === stable_fingerprint
    || record.stableFingerprint === stable_fingerprint);
  const observation = freeze({ ...findingIdentity(finding), stable_fingerprint });
  const record = freeze({
    id: existing?.id ?? stable_fingerprint,
    stable_fingerprint,
    review_round_id: existing?.review_round_id ?? existing?.reviewRoundId ?? round,
    status: existing?.status ?? 'OPEN',
    observation,
  });
  const next = existing
    ? records.map((item) => (item === existing ? record : item))
    : [...records, record];
  return freeze({
    records: freeze(next),
    record,
    disposition: existing ? 'EXISTING_RECORD' : 'NEW_RECORD',
    event: event('FINDING_UPSERTED', record),
  });
}

export class FindingLifecycleStore {
  #records = [];

  records() { return freeze([...this.#records]); }

  upsert(input) {
    const result = upsertFinding({ ...input, records: this.#records });
    this.#records = [...result.records];
    return result;
  }
}

function evidenceValue(evidence, camel, snake) { return evidence?.[camel] ?? evidence?.[snake] ?? null; }

/**
 * A closure must be based on fresh artifact evidence produced by someone other
 * than fix author. Narrative claims are intentionally not part of this check.
 */
export function verifyFindingClosure({ finding, closure, evidence }) {
  const finding_fingerprint = fingerprintFinding(finding);
  const fix_author_id = id(closure, 'fixAuthorId', 'fix_author_id');
  const reviewer_id = id(closure, 'reviewerId', 'reviewer_id');
  const evidence_reviewer_id = evidenceValue(evidence, 'reviewerId', 'reviewer_id');
  const evidence_finding_fingerprint = evidenceValue(evidence, 'findingFingerprint', 'finding_fingerprint');
  const evidence_id = evidenceValue(evidence, 'evidenceId', 'evidence_id');
  const expected_artifact_fingerprint = evidenceValue(closure, 'artifactFingerprint', 'artifact_fingerprint');
  const observed_artifact_fingerprint = evidenceValue(evidence, 'artifactFingerprint', 'artifact_fingerprint');
  if (!fix_author_id || !reviewer_id || !evidence_reviewer_id || !evidence_finding_fingerprint || !evidence_id
    || !expected_artifact_fingerprint || !observed_artifact_fingerprint) {
    return decision({ allowed: false, code: 'ARC_EVIDENCE_INSUFFICIENT', message: 'fresh independent evidence is required for finding closure', detail: { finding_fingerprint } });
  }
  if (fix_author_id === reviewer_id || fix_author_id === evidence_reviewer_id) {
    return decision({ allowed: false, code: 'ARC_SELF_CERTIFICATION', message: 'fix author cannot close own finding; fresh independent evidence is required', detail: { finding_fingerprint, fix_author_id } });
  }
  if (reviewer_id !== evidence_reviewer_id || evidence_finding_fingerprint !== finding_fingerprint
    || expected_artifact_fingerprint !== observed_artifact_fingerprint) {
    return decision({ allowed: false, code: 'ARC_BINDING_MISMATCH', message: 'closure evidence does not match finding or observed artifact', detail: { finding_fingerprint } });
  }
  const payload = freeze({ id: finding_fingerprint, finding_fingerprint, status: 'VERIFIED_CLOSED', artifact_fingerprint: observed_artifact_fingerprint, reviewer_id, evidence_id });
  return decision({ allowed: true, message: 'finding closure verified from independent artifact evidence', detail: { finding: payload, event: event('FINDING_UPSERTED', payload) } });
}

function severityValue(finding) {
  const value = finding?.severity;
  if (Number.isInteger(value) && value >= 0) return value;
  if (typeof value === 'string' && Object.hasOwn(SEVERITY, value.toLowerCase())) return SEVERITY[value.toLowerCase()];
  throw new TypeError('finding severity must be a non-negative integer or known severity');
}

/** Retain every discovery; only deterministic severity threshold decides blocking. */
export function filterFindingsByThreshold({ findings = [], threshold }) {
  if (!Number.isInteger(threshold) || threshold < 0) throw new TypeError('threshold must be a non-negative integer');
  const retained = findings.map((finding) => freeze({ ...finding, disposition: severityValue(finding) >= threshold ? 'BLOCKING' : 'INFORMATIONAL' }));
  return freeze({
    findings: freeze(retained),
    blocking: freeze(retained.filter((finding) => finding.disposition === 'BLOCKING')),
    informational: freeze(retained.filter((finding) => finding.disposition === 'INFORMATIONAL')),
    deterministic_filter: freeze({ severity_threshold: threshold }),
  });
}

/**
 * Scoped rechecks close only explicitly selected open findings. Other newly
 * observed items stay visible as deferred informational work; no full loop.
 */
export function applyScopedRecheck({ records = [], targetFindingFingerprints, target_finding_fingerprints: targetSnake, closures = [], discoveredFindings = [] }) {
  const targets = new Set(targetFindingFingerprints ?? targetSnake ?? []);
  if (!targets.size) throw new TypeError('scoped recheck requires targetFindingFingerprints');
  const closureByFingerprint = new Map(closures.map((closure) => [closure.finding_fingerprint ?? closure.findingFingerprint, closure]));
  const updated = records.map((record) => {
    const fingerprint = record.stable_fingerprint ?? record.stableFingerprint;
    if (!targets.has(fingerprint)) return record;
    const closure = closureByFingerprint.get(fingerprint);
    if (!closure?.allowed) return record;
    return freeze({ ...record, status: 'VERIFIED_CLOSED', closure_event: closure.detail?.event ?? null });
  });
  const closed = updated.filter((record) => targets.has(record.stable_fingerprint ?? record.stableFingerprint) && record.status === 'VERIFIED_CLOSED');
  const deferred = discoveredFindings.map((finding) => freeze({ ...finding, disposition: 'DEFERRED_INFORMATIONAL' }));
  return freeze({
    records: freeze(updated),
    closed: freeze(closed),
    deferred: freeze(deferred),
    events: freeze(closed.map((record) => event('FINDING_UPSERTED', { id: record.id, stable_fingerprint: record.stable_fingerprint, status: record.status }))),
  });
}

File: engine/crates/legion-arcane/src/finding_lifecycle.rs
use crate::decision::decision;
use legion_contracts::canonical_digest;
use serde_json::{json, Map, Value};
use std::sync::Mutex;

fn id(value: &Value, camel: &str, snake: &str) -> Option<String> {
    value
        .get(camel)
        .or_else(|| value.get(snake))
        .and_then(Value::as_str)
        .map(str::to_owned)
}

fn finding_identity(finding: &Value) -> Result<Map<String, Value>, String> {
    let control_id = id(finding, "controlId", "control_id")
        .or_else(|| id(finding, "ruleId", "rule_id"))
        .filter(|value| !value.is_empty())
        .ok_or_else(|| "finding controlId or ruleId is required".to_string())?;
    let subject_id = id(finding, "subjectId", "subject_id")
        .or_else(|| finding.get("file").and_then(Value::as_str).map(str::to_owned))
        .or_else(|| finding.get("path").and_then(Value::as_str).map(str::to_owned))
        .filter(|value| !value.is_empty())
        .ok_or_else(|| "finding subjectId is required".to_string())?;
    let semantic_key = id(finding, "semanticKey", "semantic_key").or_else(|| {
        id(finding, "conditionFingerprint", "condition_fingerprint")
    });
    let mut object = Map::new();
    object.insert("control_id".into(), json!(control_id));
    object.insert("subject_id".into(), json!(subject_id));
    if let Some(key) = semantic_key {
        object.insert("semantic_key".into(), json!(key));
    }
    Ok(object)
}

pub fn fingerprint_finding(finding: &Value) -> Result<String, String> {
    canonical_digest(&finding_identity(finding)?).map_err(|error| error.to_string())
}

pub fn upsert_finding(records: &[Value], finding: &Value, review_round_id: &str) -> Value {
    let stable_fingerprint = fingerprint_finding(finding).unwrap_or_default();
    let existing = records.iter().find(|record| {
        record.get("stable_fingerprint").and_then(Value::as_str) == Some(&stable_fingerprint)
            || record.get("stableFingerprint").and_then(Value::as_str) == Some(&stable_fingerprint)
    });
    let mut observation = finding_identity(finding).unwrap_or_default();
    observation.insert("stable_fingerprint".into(), json!(stable_fingerprint));
    let record = json!({
        "id": existing
            .and_then(|value| value.get("id").cloned())
            .unwrap_or(json!(stable_fingerprint)),
        "stable_fingerprint": stable_fingerprint,
        "review_round_id": existing
            .and_then(|value| value.get("review_round_id").or_else(|| value.get("reviewRoundId")).cloned())
            .unwrap_or(json!(review_round_id)),
        "status": existing
            .and_then(|value| value.get("status").cloned())
            .unwrap_or(json!("OPEN")),
        "observation": observation,
    });
    let next = if existing.is_some() {
        records
            .iter()
            .map(|item| {
                if item.get("stable_fingerprint").and_then(Value::as_str) == Some(&stable_fingerprint)
                    || item.get("stableFingerprint").and_then(Value::as_str) == Some(&stable_fingerprint)
                {
                    record.clone()
                } else {
                    item.clone()
                }
            })
            .collect::<Vec<_>>()
    } else {
        records.iter().cloned().chain([record.clone()]).collect()
    };
    json!({
        "records": next,
        "record": record,
        "disposition": if existing.is_some() { "EXISTING_RECORD" } else { "NEW_RECORD" },
        "event": {
            "event_type": "FINDING_UPSERTED",
            "payload": record,
        },
    })
}

pub struct FindingLifecycleStore {
    records: Mutex<Vec<Value>>,
}

impl FindingLifecycleStore {
    pub fn new(records: Vec<Value>) -> Self {
        Self {
            records: Mutex::new(records),
        }
    }

    pub fn records(&self) -> Vec<Value> {
        self.records.lock().unwrap().clone()
    }

    pub fn upsert(&self, finding: &Value, review_round_id: &str) -> Value {
        let records = self.records.lock().unwrap().clone();
        let result = upsert_finding(&records, finding, review_round_id);
        if let Some(next) = result.get("records").and_then(Value::as_array) {
            *self.records.lock().unwrap() = next.clone();
        }
        result
    }
}

fn evidence_value(evidence: &Value, camel: &str, snake: &str) -> Option<String> {
    evidence
        .get(camel)
        .or_else(|| evidence.get(snake))
        .and_then(Value::as_str)
        .map(str::to_owned)
}

pub fn verify_finding_closure(input: &Value) -> Value {
    let finding = input.get("finding").unwrap_or(&Value::Null);
    let closure = input.get("closure").unwrap_or(&Value::Null);
    let evidence = input.get("evidence").unwrap_or(&Value::Null);
    let finding_fingerprint = fingerprint_finding(finding).unwrap_or_default();
    let fix_author_id = id(closure, "fixAuthorId", "fix_author_id");
    let reviewer_id = id(closure, "reviewerId", "reviewer_id");
    let evidence_reviewer_id = evidence_value(evidence, "reviewerId", "reviewer_id");
    let evidence_finding_fingerprint =
        evidence_value(evidence, "findingFingerprint", "finding_fingerprint");
    let evidence_id = evidence_value(evidence, "evidenceId", "evidence_id");
    let expected_artifact_fingerprint =
        evidence_value(closure, "artifactFingerprint", "artifact_fingerprint");
    let observed_artifact_fingerprint =
        evidence_value(evidence, "artifactFingerprint", "artifact_fingerprint");
    if fix_author_id.is_none()
        || reviewer_id.is_none()
        || evidence_reviewer_id.is_none()
        || evidence_finding_fingerprint.is_none()
        || evidence_id.is_none()
        || expected_artifact_fingerprint.is_none()
        || observed_artifact_fingerprint.is_none()
    {
        return decision(
            false,
            Some("ARC_EVIDENCE_INSUFFICIENT"),
            Some("fresh independent evidence is required for finding closure"),
            json!({ "finding_fingerprint": finding_fingerprint }),
        );
    }
    if fix_author_id == reviewer_id || fix_author_id.as_ref() == evidence_reviewer_id.as_ref() {
        return decision(
            false,
            Some("ARC_SELF_CERTIFICATION"),
            Some("fix author cannot close own finding; fresh independent evidence is required"),
            json!({
                "finding_fingerprint": finding_fingerprint,
                "fix_author_id": fix_author_id,
            }),
        );
    }
    if reviewer_id.as_ref() != evidence_reviewer_id.as_ref()
        || evidence_finding_fingerprint.as_deref() != Some(finding_fingerprint.as_str())
        || expected_artifact_fingerprint.as_ref() != observed_artifact_fingerprint.as_ref()
    {
        return decision(
            false,
            Some("ARC_BINDING_MISMATCH"),
            Some("closure evidence does not match finding or observed artifact"),
            json!({ "finding_fingerprint": finding_fingerprint }),
        );
    }
    let payload = json!({
        "id": finding_fingerprint,
        "finding_fingerprint": finding_fingerprint,
        "status": "VERIFIED_CLOSED",
        "artifact_fingerprint": observed_artifact_fingerprint,
        "reviewer_id": reviewer_id,
        "evidence_id": evidence_id,
    });
    decision(
        true,
        None,
        Some("finding closure verified from independent artifact evidence"),
        json!({
            "finding": payload,
            "event": {
                "event_type": "FINDING_UPSERTED",
                "payload": payload,
            },
        }),
    )
}

pub fn filter_findings_by_threshold(input: &Value) -> Result<Value, String> {
    let threshold = input
        .get("threshold")
        .and_then(Value::as_i64)
        .filter(|value| *value >= 0)
        .ok_or_else(|| "threshold must be a non-negative integer".to_string())?;
    let findings = input
        .get("findings")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    let retained = findings
        .into_iter()
        .map(|finding| {
            let severity = severity_value(&finding)?;
            let disposition = if severity >= threshold {
                "BLOCKING"
            } else {
                "INFORMATIONAL"
            };
            let mut object = finding.as_object().cloned().unwrap_or_default();
            object.insert("disposition".into(), json!(disposition));
            Ok(Value::Object(object))
        })
        .collect::<Result<Vec<_>, String>>()?;
    let blocking = retained
        .iter()
        .filter(|finding| finding.get("disposition") == Some(&json!("BLOCKING")))
        .cloned()
        .collect::<Vec<_>>();
    let informational = retained
        .iter()
        .filter(|finding| finding.get("disposition") == Some(&json!("INFORMATIONAL")))
        .cloned()
        .collect::<Vec<_>>();
    Ok(json!({
        "findings": retained,
        "blocking": blocking,
        "informational": informational,
        "deterministic_filter": { "severity_threshold": threshold },
    }))
}

fn severity_value(finding: &Value) -> Result<i64, String> {
    if let Some(value) = finding.get("severity").and_then(Value::as_i64) {
        if value >= 0 {
            return Ok(value);
        }
    }
    if let Some(text) = finding.get("severity").and_then(Value::as_str) {
        let mapped = match text.to_ascii_lowercase().as_str() {
            "informational" => 0,
            "low" => 1,
            "medium" => 2,
            "high" => 3,
            "critical" => 4,
            _ => return Err("finding severity must be a non-negative integer or known severity".into()),
        };
        return Ok(mapped);
    }
    Err("finding severity must be a non-negative integer or known severity".into())
}

pub fn apply_scoped_recheck(input: &Value) -> Result<Value, String> {
    let targets = input
        .get("targetFindingFingerprints")
        .or_else(|| input.get("target_finding_fingerprints"))
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    if targets.is_empty() {
        return Err("scoped recheck requires targetFindingFingerprints".into());
    }
    let target_set = targets
        .iter()
        .filter_map(Value::as_str)
        .collect::<std::collections::HashSet<_>>();
    let records = input
        .get("records")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    let closures = input
        .get("closures")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    let closure_by_fingerprint = closures
        .iter()
        .filter_map(|closure| {
            let fingerprint = closure
                .get("finding_fingerprint")
                .or_else(|| closure.get("findingFingerprint"))
                .and_then(Value::as_str)?;
            Some((fingerprint, closure))
        })
        .collect::<std::collections::HashMap<_, _>>();
    let updated = records
        .iter()
        .map(|record| {
            let fingerprint = record
                .get("stable_fingerprint")
                .or_else(|| record.get("stableFingerprint"))
                .and_then(Value::as_str)
                .unwrap_or_default();
            if !target_set.contains(fingerprint) {
                return record.clone();
            }
            let closure = closure_by_fingerprint.get(fingerprint);
            if closure.and_then(|value| value.get("allowed")).and_then(Value::as_bool) != Some(true) {
                return record.clone();
            }
            let mut object = record.as_object().cloned().unwrap_or_default();
            object.insert("status".into(), json!("VERIFIED_CLOSED"));
            if let Some(closure) = closure {
                object.insert(
                    "closure_event".into(),
                    closure
                        .pointer("/detail/event")
                        .cloned()
                        .unwrap_or(Value::Null),
                );
            }
            Value::Object(object)
        })
        .collect::<Vec<_>>();
    let closed = updated
        .iter()
        .filter(|record| {
            let fingerprint = record
                .get("stable_fingerprint")
                .or_else(|| record.get("stableFingerprint"))
                .and_then(Value::as_str)
                .unwrap_or_default();
            target_set.contains(fingerprint) && record.get("status") == Some(&json!("VERIFIED_CLOSED"))
        })
        .cloned()
        .collect::<Vec<_>>();
    let deferred = input
        .get("discoveredFindings")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default()
        .into_iter()
        .map(|finding| {
            let mut object = finding.as_object().cloned().unwrap_or_default();
            object.insert("disposition".into(), json!("DEFERRED_INFORMATIONAL"));
            Value::Object(object)
        })
        .collect::<Vec<_>>();
    Ok(json!({
        "records": updated,
        "closed": closed,
        "deferred": deferred,
        "events": closed.iter().map(|record| json!({
            "event_type": "FINDING_UPSERTED",
            "payload": {
                "id": record.get("id").cloned().unwrap_or(Value::Null),
                "stable_fingerprint": record.get("stable_fingerprint").or_else(|| record.get("stableFingerprint")).cloned().unwrap_or(Value::Null),
                "status": record.get("status").cloned().unwrap_or(Value::Null),
            }
        })).collect::<Vec<_>>(),
    }))
}
